### PR TITLE
fix(bindnode): tuple struct iterator should handle absent fields properly

### PR DESCRIPTION
**Pull request by rvagg**
Fixes: https://github.com/ipld/go-ipld-prime/issues/367

### CODE REVIEW

File: node/bindnode/repr.go [hunk 1]
```diff
@@ -320,7 +322,7 @@ _skipAbsent:
 	if err != nil {
 		return 0, nil, err
 	}
-	if value.IsAbsent() {
+	if w.nextIndex > w.reprEnd {
```
Comment (rvagg) [inline node/bindnode/repr.go:325]: Is it possible to be in a situation where `value.IsAbsent()` and we're not at the end amongst trailing absents? e.g. `[absent,present,present,absent,absent]`? Is there a use of this iterator where that can happen, perhaps by constructing this from the typed form and then list iterating on the repr form?

**Comment by rvagg** [comment]
btw, I could go either way on the tuple optionals, I've always thought that it was reasonable to allow trailing optionals but it does complicate matters somewhat for marginal gain and increased ambiguity. Maybe we should rip that support out (for now, or for good?)

File: node/bindnode/repr.go (hunk 1 above)
Comment (mvdan) [reply node/bindnode/repr.go:325]: Absents can only be at the end, as per the undocumented spec :) See https://github.com/ipld/go-ipld-prime/blob/e39d20bf18dcc21a50ef2005bae6de11e5c311e0/schema/gen/go/genStructReprTuple.go#L13.

Arguably this needs to be properly documented in the specs, which is why I'll raise an issue elsewhere. But in short: we should support trailingn optionals in bindnode, to at least match codegen in terms of features.

**Review by mvdan — APPROVED** [review]

**Comment by mvdan** [comment]
TL;DR fine to merge as is, looks great!

File: node/bindnode/repr.go [hunk 2]
```diff
@@ -348,11 +350,7 @@ func (w *_nodeRepr) Length() int64 {
 	case schema.StructRepresentation_Map:
 		return w.lengthMinusAbsents()
 	case schema.StructRepresentation_Tuple:
-		// TODO: presumably, tuples should disallow optional fields.
-		// But the following tests disagrees:
-		// https://github.com/ipld/go-ipld-prime/blob/297518c911326ef7f44800dc32a923e7b084fccc/node/tests/schemaStructReprTuple.go#L24-L32
-		// return int64(len(w.schemaType.(*schema.TypeStruct).Fields()))
```
[the remark this answers is not in the document]
Comment (mvdan) [reply node/bindnode/repr.go:354]: Now https://github.com/ipld/go-ipld-prime/issues/369.
